Jira Ticket Task

zhen0 opened the pull request:
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
- Adds a Jira notifier that creates and assigns a jira issue if a task enters a specific state (probably failed but I've left it open and can be decided at time of use)


## Why is this PR important?
- Requested in issue Jira Ticket Task #1836

Creating as a draft as I realize this needs tests, changelog update and possibly some docs

cicdw commented on src/prefect/utilities/notifications.py:
```diff
@@ -279,5 +292,83 @@ def add(x, y):
+    serverURL = cast(str, prefect.client.Secret("JIRASERVER").get())
+
+    if not projectName:
+        projectName = cast(str, prefect.client.Secret("JIRAPROJECT").get())
```
Two thoughts on the use of secrets here:
- outside of the scope of this PR, it would be nice to find a way around _requiring_ Prefect Cloud Secrets
- since you require 4 distinct secrets, you might consider refactoring this to one `JIRA_CREDS` secret which is a JSON payload (a dictionary in Python) with these four keys; might make the setting of the secret easier on the user (that is also just my own personal preference - open to push back)

dylanbhughes replied:
I like Chris's suggestion here

cicdw commented on src/prefect/utilities/notifications.py:
```diff
@@ -279,5 +292,83 @@ def add(x, y):
+    assignee: str = -1
+) -> "prefect.engine.state.State":
+    """
+    Slack state change handler; requires having the Prefect slack app installed.
```
This whole docstring is still the slack docstring

cicdw commented on src/prefect/utilities/notifications.py:
```diff
@@ -152,6 +157,13 @@ def slack_message_formatter(
 
 
+def jira_message_formatter(tracked_obj: TrackedObjectType, state: "prefect.engine.state.State") -> str:
+    time = datetime.now()
```
This call will create a timestamp in the local timezone; you might want to consider using `datetime.utcnow()` instead

cicdw commented on src/prefect/utilities/notifications.py:
```diff
@@ -279,5 +292,83 @@ def add(x, y):
+
+
+@curry
+def jira_notifier(
```
In terms of docs, we'll also want to add this to the `docs/outline.toml` file

aroder commented:
@zhen0 I like the idea of being able to pass an assignee, and I think that we will need more fields than that. We have some custom Jira fields that we would want to pass in, that allow us to use Jira to configure automatic routing and assignment of newly created tickets. Have you found a way to pass in a collection of custom fields? 

Our use case is Jira Service Desk, but I have not had a chance to review and compare the APIs between Jira Software and Jira Service Desk. They may be the same API.

From past experience, these kinds of automatic ticket creation triggers can result in thousands of tickets being created. This could be handled by the developer if careful thought is put into when a ticket is generated--but often there are emergent cases not thought of during design. The pattern I've seen that handles this is for the state handler to recognize a case when lots of duplicate tickets are about to be created, and do something like create 1 ticket with a note that the event occurred N number of times. This definitely is not a priority for us, just some feedback as you design this.

zhen0 commented:
@cicdw - I think this is ready for another review.  I've incorporated your requested changes (one jira  creds secret, time in utc, added docs and tests.). I've also added a bit more flexibility for tailoring the call to jira which should allow adding different issue types (which means the state handler can work for jira service desk as well as standard jira) and custom fields.

joshmeek commented:
@zhen0 Do you happen to have an image of what this notification looks like in Jira?

zhen0 commented:
<img width="1020" alt="Screen Shot 2020-02-28 at 11 52 06 AM" src="https://user-images.githubusercontent.com/40272060/75568177-cba0fc80-5a20-11ea-8568-8c6bb61544bb.png">

Here you go @joshmeek

zhen0 commented:
And from Jira Service Desk:

<img width="1042" alt="Screen Shot 2020-02-28 at 11 57 29 AM" src="https://user-images.githubusercontent.com/40272060/75568586-8e893a00-5a21-11ea-927f-80b7c662ec23.png">

cicdw commented on src/prefect/utilities/jira_notification.py:
```diff
@@ -0,0 +1,121 @@
+if TYPE_CHECKING:
+    import prefect.engine.state
+    import prefect.client
+    from prefect import Flow, Task
```
One of our CI checks is failing because of these unused imports; I recommend removing this line and replacing

```python
TrackedObjectType = Union[prefect.Flow, prefect.Task]
```

below

zhen0 replied:
A ha!  Thank you!

zhen0 commented on docs/outline.toml:
```diff
@@ -449,6 +449,11 @@ title = "Notifications and Callback Tools"
 module = "prefect.utilities.notifications"
 functions = ["callback_factory", "slack_notifier", "gmail_notifier"]
 
+[pages.utilities.jira_notification]
```
As jira is not imported as standard, it needs to be imported as an extra.  We thought moving it to its own page would allow us to raise a warning if jira is not already installed for the jira notifier without affecting the other notifiers.

cicdw replied:
Gotcha; so whether it requires an extra or not actually doesn't matter for the auto-generated documentation -- I'm able to patch any missing imports so that I can still extract the relevant docstrings.  If you put it in the same page as the others, you'd need to add a 
```python
try:
    from .jira import jira_notifier # might not be 100%
except ImportError:
    pass
```
into `utilities.notifications.__init__.py`

Up to you though, if you want it on its own documentation page that's OK

joshmeek commented on src/prefect/utilities/jira_notification.py:
```diff
@@ -0,0 +1,120 @@
+    assignee: str = "-1",
+) -> "prefect.engine.state.State":
+    """
+    Jira Notifier requires a Jira account and API token.  They API token can be created at: https://id.atlassian.com/manage/api-tokens 
```
```suggestion
    Jira Notifier requires a Jira account and API token.  The API token can be created at: https://id.atlassian.com/manage/api-tokens 
```

joshmeek commented on src/prefect/utilities/jira_notification.py:
```diff
@@ -0,0 +1,123 @@
+    Jira Notifier creates a new ticket with the information about the task or flow it is bound to when that task or flow is in a specific state. 
+    (For example it will create a ticket to tell you that the flow you set it on is in a failed state.)  
+    You should use the options dictionary to set the project name and issue type.  
+    You can use the "assignee" argument to assign that ticket to a specific member of your team.
```
```suggestion
    Jira Notifier requires a Jira account and API token.  The API token can be created at: https://id.atlassian.com/manage/api-tokens 
    The Jira account username ('JIRAUSER'), API token ('JIRATOKEN') should be set as part of a 'JIRASECRETS' object in Prefect Secrets. 

    An example 'JIRASECRETS' object looks like this:
    ```
    JIRASECRETS = { JIRATOKEN = "XXXXXXXXX", JIRAUSER = "xxxxx@yyy.com", JIRASERVER = "https://???.atlassian.net", JIRAPROJECT = "TEST" }
    ```

    The server URL can be set as part of the 'JIRASECRETS' object ('JIRASERVER') or passed to the jira notifier state handler as the "server_URL" argument.
    Jira Notifier works as a standalone state handler, or can be called from within a custom
    state handler.  This function is curried meaning that it can be called multiple times to partially bind any keyword arguments (see example below).
    Jira Notifier creates a new ticket with the information about the task or flow it is bound to when that task or flow is in a specific state. 
    (For example it will create a ticket to tell you that the flow you set it on is in a failed state.)  
    You should use the options dictionary to set the project name and issue type.  
    You can use the "assignee" argument to assign that ticket to a specific member of your team.
```

cicdw commented:
OK so I got this working so that the new jira notifier can be documented alongside the other notifiers and bypasses import issues.  You can see my changes here: https://github.com/PrefectHQ/prefect/pull/1861/commits/7ff4d8f17476da8802495f7b76d84e728f60cd74

Note that this change allows us to document classes / functions from un-imported submodules alongside other classes / functions
